engine/e_bluefield.c

```diff
@@ -829,6 +924,743 @@ static int engine_pka_dh_finish(DH *dh)
+
+    if (ossl_pkey == NULL || ossl_peerkey == NULL)
+    {
+        printf("ERROR: engine_pka_X25519_derive: Keys are not set\n");
+        return 0;
+    }
+
+    pkey = EVP_PKEY_get0(ossl_pkey);
+    peerkey = EVP_PKEY_get0(ossl_peerkey);
+
+    if (engine_pka_keypair_invalid(pkey, NID_X25519, 1))
+    {
+        printf("ERROR: engine_pka_X25519_derive: pkey is invalid\n");
+        return 0;
+    }
+
+    if (engine_pka_keypair_invalid(peerkey, NID_X25519, 1))
+    {
+        printf("ERROR: engine_pka_X25519_derive: peerkey is invalid\n");
+        return 0;
+    }
+    *keylen = PKA_25519_PUBKEY_SIZE;
+
+    if (key != NULL &&
+        !pka_mont_25519_mult(key, &peerkey->public_key,
+                             &pkey->private_key))
+        return 0;
+
+    return 1;
+}
+
+/* Montgomery Curve 448 PKEY operations */
+
+static int engine_pka_X448_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey)
+{
+    ENGINE_PKA_KEYPAIR *kpair = NULL;
+
+    kpair = engine_pka_keypair_new(NID_X448, 0, PKA_448_PRIKEY_SIZE);
+    if (!kpair)
+    {
+        printf("ERROR: engine_pka_X448_keygen: Invalid keypair\n");
+        goto err;
+    }
+
+    if (!engine_pka_get_random_bytes(kpair->private_key.buf_ptr,
+                                     PKA_448_PRIKEY_SIZE))
+    {
+        printf("ERROR: engine_pka_X448_keygen: Failed to get random bytes\n");
+        goto err;
+    }
+
+    if (!pka_mont_448_derive_pubkey(kpair->public_key.buf_ptr,
+                                    &kpair->private_key))
+    {
+        printf("ERROR: engine_pka_X448_keygen: "
+               "Failed to generate public key\n");
+        goto err;
+    }
+
+    EVP_PKEY_assign(pkey, NID_X448, kpair);
+    return 1;
+
+err:
+    if (kpair)
+        engine_pka_keypair_free(kpair);
+    return 0;
+}
+
+static int
+engine_pka_X448_derive(EVP_PKEY_CTX *ctx, unsigned char *key, size_t *keylen)
+{
+    EVP_PKEY *ossl_pkey, *ossl_peerkey;
+    ENGINE_PKA_KEYPAIR *pkey, *peerkey;
+
+    ossl_pkey = EVP_PKEY_CTX_get0_pkey(ctx);
+    ossl_peerkey = EVP_PKEY_CTX_get0_peerkey(ctx);
+
+    if (ossl_pkey == NULL || ossl_peerkey == NULL)
+    {
+        printf("ERROR: engine_pka_X448_derive: Keys are not set\n");
+        return 0;
+    }
+
+    pkey = EVP_PKEY_get0(ossl_pkey);
+    peerkey = EVP_PKEY_get0(ossl_peerkey);
+
+    if (engine_pka_keypair_invalid(pkey, NID_X448, 1))
+    {
+        printf("ERROR: engine_pka_X448_derive: pkey is invalid\n");
+        return 0;
+    }
+
+    if (engine_pka_keypair_invalid(peerkey, NID_X448, 1))
+    {
+        printf("ERROR: engine_pka_X448_derive: peerkey is invalid\n");
+        return 0;
+    }
+    *keylen = PKA_448_PUBKEY_SIZE;
+
+    if (key != NULL &&
+        !pka_mont_448_mult(key, &peerkey->public_key,
+                           &pkey->private_key))
+        return 0;
+
+    return 1;
+}
+
+static int engine_pka_X_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
+{
+    if (type == EVP_PKEY_CTRL_PEER_KEY)
+        return 1;
+
+    return -2;
```

Comment:
Question: why this error code is <0? (unlike the rest)

Reply:
This function is hooked into OpenSSL using `EVP_PKEY_meth_set_ctrl()` below. OpenSSL requires it to return -2 specifically to notify operation is not supported. Please see **RETURN VALUES** section here: https://www.openssl.org/docs/man1.1.1/man3/EVP_PKEY_CTX_ctrl.html

Reply:
Thanks